vllm/model_executor/models/jina_embeddings_v4.py

```diff
@@ -0,0 +1,327 @@
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+import time
+from array import array
+from collections.abc import Iterable
+from typing import Optional, Union
+
+import torch
+import torch.nn.functional as F
+from typing_extensions import assert_never
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.model_executor.layers.pooler import (HAS_TRITON, Pooler, PoolingTask,
+                                               PoolingType, build_output,
+                                               extract_vision_tokens_kernel,
+                                               get_prompt_lens)
+# yapf: disable
+from vllm.model_executor.pooling_metadata import (
+    PoolingMetadata as V0PoolingMetadata)
+# yapf: enable
+from vllm.multimodal import MULTIMODAL_REGISTRY
+from vllm.pooling_params import PoolingParams
+from vllm.sequence import PoolerOutput
+from vllm.v1.pool.metadata import PoolingMetadata as V1PoolingMetadata
+
+from .interfaces import SupportsCrossEncoding, SupportsMultiModal
+from .qwen2_vl import (Qwen2VLDummyInputsBuilder,
+                       Qwen2VLForConditionalGeneration,
+                       Qwen2VLMultiModalProcessor, Qwen2VLProcessingInfo)
+from .utils import AutoWeightsLoader, WeightsMapper, maybe_prefix
+
+logger = init_logger(__name__)
+
+# Vision token IDs for Jina V4
+VISION_START_TOKEN_ID = 151652
+VISION_END_TOKEN_ID = 151653
+
+PoolingMetadata = Union[V0PoolingMetadata, V1PoolingMetadata]
+
+
+class JinaVLPooler(Pooler):
+    """Vision-aware pooler for Jina V4 with special vision token handling."""
+
+    def __init__(self,
+                 vllm_config: VllmConfig,
+                 pooling_backend: str = "pytorch"):
+        super().__init__()
+        self.hidden_size = vllm_config.model_config.hf_config.hidden_size
+        self.pooling_backend = pooling_backend
+        self.observability_config = vllm_config.observability_config
+
+        # Pre-compute vision token IDs tensor for efficient checking
+        self.vision_token_ids = torch.tensor(
+            [VISION_START_TOKEN_ID, VISION_END_TOKEN_ID], dtype=torch.long)
+
+        # Performance tracking
+        self._pooling_time_ms = 0.0
+        self._pooling_count = 0
+
+        # Initialize base pooler for fallback
+        pooler_config = vllm_config.model_config.pooler_config
+        self._base_pooler = Pooler.from_config_with_defaults(
+            pooler_config,
+            pooling_type=PoolingType.MEAN,
+            normalize=True,
+            softmax=False)
+
+    def get_pooling_params(self, task: PoolingTask) -> Optional[PoolingParams]:
+        """Return pooling params for embedding task."""
+        if task == "embed" or task == "encode":
+            return PoolingParams(logits_processing_needs_token_ids=True)
+
+        # The equalities are split up to keep mypy happy
+        if task == "classify" or task == "score":
+            return None
+
+        assert_never(task)
```

Comment:
Can you merge from main and then apply this update? (Need to update the imports accordingly as well)

```suggestion
    def get_pooling_updates(
        self,
        task: PoolingTask,
    ) -> Optional[PoolingParamsUpdate]:
        # The equalities are split up to keep mypy happy
        if task == "encode" or task == "embed":
            return PoolingParamsUpdate(requires_token_ids=True)

        if task == "classify" or task == "score":
            return None

        assert_never(task)
```